Fix broken expectation in Safari for serializing invalid dates.

BobChao87 opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Bug fix.

**What is the current behavior? (You can also link to an open issue here)**

Fixes #13415. Current behaviour is that Safari throws a render error instead of correctly handling Invalid Dates (such as `new Date(1/0)`).

**What is the new behavior (if this is a feature change)?**

New behaviour is that Safari performs the same as other browsers, by serializing the invalid date or any `RangeError`) to `'null'`.

**Does this PR introduce a breaking change?**

No.

**Please check if the PR fulfills these requirements**
- [Yes] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [Yes] Tests for the changes have been added (for bug fixes / features)
- [Unnecessary] Docs have been added / updated (for bug fixes / features)

**Other information**:

Added test to prevent regression. No docs need updating since no behaviour changed and no new features were added.

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

BobChao87 committed:
Fix broken expection in Safari for serializing invalid dates. All browsers should consistently serialize to 'null'.

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

BobChao87 committed:
Merge branch 'master' of https://github.com/BobChao87/angular.js

Error with commit comment for Angular standards.

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

BobChao87 commented:
Sorry about the spam of commits. I'm not sure what I'm doing with git. It's all new to me.

I'm not sure what to do about the CLA issue since @googlebot claims it found a CLA for me, but not for the committing party, but I am the only committing party.

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

petebacondarwin commented:
The build is green now (it was just Travis being flaky)
I will find out what to do about the CLA

gkalpak commented:
The CLA issue is probably because at some point you merged master into your branch.
You could try to rebase removing that commit (and optionally squach your other commits into one).

gkalpak commented on src/Angular.js:
```diff
@@ -1195,7 +1195,12 @@ function toJson(obj, pretty) {
+    return JSON.stringify(obj, toJsonReplacer, pretty);
+  } catch (RangeError) {
+    // In Safari, 'Invalid Date's throw a range error. 'null' is returned in all other browsers.
+    return 'null';
```
This means that the whole Object will be serialized as `"null"` even if one sub-property is an invalid date.
This is inconsistent with other browsers and doesn't feel right either.

I think it is more appropriate to handle this inside the `toJsonReplacer()` function.

gkalpak commented on test/AngularSpec.js:
```diff
@@ -2058,6 +2058,10 @@ describe('angular', function() {
     });
+
+    it('should serialize invalid dates to null', function() {
+      expect(toJson(new Date(1 / 0))).toEqual('null');
```
To cover more cases, put the invalid date inside an object (e.g. `toJson({when: new Date(1 / 0)})`).

IgorMinar commented:
I was able to manually verify the CLA signature. I'll file a bug against the bot owners to support verification for users with hidden email addresses.

petebacondarwin commented:
@BobChao87 - we have sorted out the CLA problems; can you address the most recent comments then we can get this merged.

BobChao87 committed:
fix(toJson): correct serialization of invalid dates

Fix for error thrown by Safari when an invalid date is serialized. Use 'null' always.

Closes #13415

BobChao87 commented:
That should do it. The replacer was a good idea, I was clearly too focused on the exact example provided. Thanks @IgorMinar for the CLA fixes.

BobChao87 commented:
Checks for the invalid date worked outside. They don't seem to work inside the `toJsonReplacer` because the date isn't passed in as a Date. I'll try to figure something out.

gkalpak commented:
Indeed. I couldn't figure something out. Maybe this problem is not solveable (at least not without having to traverse the whole object which is obviously out of question).

Or maybe you have better luck figuring something out...

Narretz commented:
I think as long as the issue can't be fixed satisfactorily, we'll put this into the regular 1.5 milestone.

petebacondarwin commented:
Here is a hack that fixes this problem:

```js
var DatetoJSON = Date.prototype.toJSON;
Date.prototype.toJSON = function() {
  try {
    return DatetoJSON.call(this);
  } catch(e) {
    if (e instanceof RangeError) {
      return null;
    }
    throw e;
  }
};
```

The question is how best to sniff for Safari and only do it for that browser?

petebacondarwin commented:
But since this sort of changing built in global objects is not nice, I think we cannot do this in Angular.
It is a perfectly reasonable workaround for your own apps though...

petebacondarwin commented:
I'll document this as a known issue, with this workaround over the weekend.
But I think we will have to leave this as it is and close this PR and the related issue.
